### ISSUE TITLE
Failed to apply XNNPACK after quantization aware training in Tensorflow 2.7

### ISSUE THREAD

**Issue by Janus-Shiau** [google/XNNPACK#3693]
We use quantization-aware training to train our models in tensorflow 2.7.0, and then deploy as `tf.int8` TFLite models. 

The model can be converted successful, and fully functional on Qualcomm's CPU and Hexagon. But it run into errors when using XNNPACK delegate. Actually, XNNPACK is not always failed. It can run normally in some case:

1. If we deploy the model directly without training, it works.
2. Basically we training the model for 40k steps. if we traiqualcommn it shortly, e.g. only 25k, it works.
3. The backbone of the model is HRNet, If we used another architecture, e.g. Hourglass, it works.

We have been haunted by this strange thing for a month, please kindly let us know what's wrong.

Thank you for your time.

### 1. System information

- OS Platform and Distribution: Ubuntu 18.04
- TensorFlow installation (pip package): 2.7.0
- TensorFlow library (version, if pip package or github SHA, if built from source): 2.7.0

### 2. Code

The model has two inputs, where the range of first one is [0, 1.0] and the second one is [-1.0, 1.0].

```python
converter = tf.compat.v1.lite.TFLiteConverter.from_frozen_graph(
    os.path.join(dir_name, "model.pb"), inputs, outputs
)
converter.target_spec.supported_ops = [
    tf.lite.OpsSet.TFLITE_BUILTINS,  tf.lite.OpsSet.SELECT_TF_OPS,
]

converter.inference_type = tf.int8
converter.default_ranges_stats = (-128, 127)
converter.quantized_input_stats = {
    inputs[0]: {128, 128},
    **({inputs[1]: {0, 128}} if len(inputs) == 2 else {}),
}
output_path = os.path.join(output_path, name + "_quant.tflite")

tflite_model = converter.convert()
open(output_path, "wb").write(tflite_model)
```

### 3. Failure information

We test the model with tflite benchmark with xnnpack enabled, and found the following error messages:

```bash
STARTING!
Log parameter values verbosely: [0]
Min num runs: [250]
Num threads: [3]
Graph: [/data/debug/model_quant.tflite]
Enable op profiling: [1]
#threads used for CPU inference: [3]
Use xnnpack: [1]
Loaded model /data/debug/model_quant.tflite
INFO: Initialized TensorFlow Lite runtime.
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
XNNPACK delegate created.
INFO: Replacing 132 node(s) with delegate (TfLiteXNNPackDelegate) node, yielding 10 partitions.
ERROR: failed to create XNNPACK Value for tensor 103
ERROR: Node number 147 (TfLiteXNNPackDelegate) failed to prepare.

ERROR: Restored original execution plan after delegate application failure.
Failed to apply XNNPACK delegate.
Benchmarking failed.
```

**Comment by Maratyszcza** [comment]
The issue is most certainly in the XNNPack delegate for TensorFlow Lite and not in the XNNPack library itself.

Please re-build TensorFlow Lite with `--define xnn_log_level=info` and re-run the benchmark. This will provide more detailed logging about what has failed.

**Comment by Janus-Shiau** [comment]
I rebuild the library on Tensorflow v2.7.0 using docker environment, and use the following command for bazel:

```
bazel build -c opt --config=android_arm64 --define tflite_with_xnnpack=true --define xnn_enable_qs8=true --define xnn_log_level=info //tensorflow/lite/tools/benchmark:benchmark_mode
```

Then copy and push `benchmark_model` into the device with adb:

```
cp bazel-bin/tensorflow/lite/tools/benchmark/benchmark_model .
adb push benchmark_model /data/debug/
```

Finally run test again:

```
adb shell /data/debug/benchmark_model --graph=/data/debug/model_quant.tflite --enable_op_profiling=0 --use_xnnpack=1
```

But there is nothing addiction information pop out.

```bash
STARTING!
Log parameter values verbosely: [0]
Min num runs: [250]
Num threads: [3]
Graph: [/data/debug/model_quant.tflite]
Enable op profiling: [0]
#threads used for CPU inference: [3]
Use Hexagon: [0]
Use xnnpack: [1]
Loaded model /data/debug/model_quant.tflite
INFO: Initialized TensorFlow Lite runtime.
INFO: XNNPACK delegate created.
Created TensorFlow Lite XNNPACK delegate for CPU.
INFO: Replacing 133 node(s) with delegate (TfLiteXNNPackDelegate) node, yielding 10 partitions.
ERROR: failed to create XNNPACK Value for tensor 103
ERROR: Node number 147 (TfLiteXNNPackDelegate) failed to prepare.

ERROR: Restored original execution plan after delegate application failure.
Failed to apply XNNPACK delegate.
Benchmarking failed.
```

Did I miss anything here?

**Comment by Maratyszcza** [comment]
If you run on Android, also add `--define xnn_log_to_stdio=true`. Otherwise, the information is logged to `adblog`.

**Comment by Janus-Shiau** [comment]
I have check the `adblog` with `adb logcat` but didn't find any addiction information there.

Then I just followed your instruction, cleaned the bazel cache and rebuilded the `benchmark_model`:

```console
bazel build -c opt --config=android_arm64 --define tflite_with_xnnpack=true \
    --define xnn_enable_qs8=true --define xnn_log_level=info --define xnn_log_to_stdio=true \  
    //tensorflow/lite/tools/benchmark:benchmark_model
```

And use the same command above, but still get nothing more on either `stdio` or `adblog`.

**Comment by ngzhian** [comment]
`--define xnn_log_level=debug` will give more logging, maybe you can try that too.

**Comment by Janus-Shiau** [comment]
We also try the flag `debug` before,and now again, but still get nothing new.

**Comment by Maratyszcza** [comment]
Could you show what tensor 103 looks like in netron?

**Comment by Janus-Shiau** [comment]
The node 103 is the bias of `Conv2D` layers, including the batch normalization layer folding.

![image](https://user-images.githubusercontent.com/22385182/199665737-c9a235c8-0bc9-4f4a-8d31-d725aaaff0df.png)

Some value of that bias is listed as for your reference.

```
type: int32[16384]
quantization: 
  3.820020133549207e-10 * q
  5.148818593703197e-10 * q
  6.663699592124317e-10 * q
  5.802464064785795e-10 * q
  4.573305623090107e-10 * q
  ....
[
    -521663123,
    -462378818,
    -454766690,
    -387756887,
    -597149483,
    ....
]
```

And 147 is the ReLU output of the `DepthwiseConv2D` layer.

![image](https://user-images.githubusercontent.com/22385182/199667182-f7c756cf-9e28-4658-9776-337329bd9c92.png)


Thank you for your time and comments.

**Comment by Janus-Shiau** [comment]
I also provide a tester model for you, which contains similar the above error:

```
STARTING!
Log parameter values verbosely: [0]
Min num runs: [250]
Num threads: [3]
Graph: [/data/debug/model_quant.tflite]
Enable op profiling: [1]
#threads used for CPU inference: [3]
Use xnnpack: [1]
Loaded model /data/debug/model_quant.tflite
INFO: Initialized TensorFlow Lite runtime.
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
XNNPACK delegate created.
INFO: Replacing 36 node(s) with delegate (TfLiteXNNPackDelegate) node, yielding 6 partitions.
ERROR: failed to create XNNPACK Value for tensor 31
ERROR: Node number 42 (TfLiteXNNPackDelegate) failed to prepare.

ERROR: Restored original execution plan after delegate application failure.
Failed to apply XNNPACK delegate.
Benchmarking failed.
```

The model download link:
https://drive.google.com/file/d/19XB3XwbvfhmsvFJ_SKuDQi_QgN5L--2t/view?usp=share_link

**Comment by Maratyszcza** [comment]
Here's what I get with logging enabled: `Error in XNNPACK: failed to create Channelwise Quantized Dense Tensor value with 7.222319e-39 scale in channel #6: scale must be finite, normalized, and positive`. I.e. quantized inference failed because one of the per-channel scales in the bias vector is practically zero.

**Comment by Janus-Shiau** [comment]
I found there might be two error here. One is ***"ERROR: failed to create XNNPACK Value for tensor 31"***, this might due to the reason your provided. And it can solved this by forcing that layer keep as `float32`. But the second one, ***"ERROR: Node number 42 (TfLiteXNNPackDelegate) failed to prepare."***, is still unknown.

And I'm also quite confused about why Hexagon and standard CPU library can still execute this quantized model.

Besides, I'm very appreciate your works about running debug message for us. Because I'm not able to get these message even with `--define xnn_log_level=debug` when building  `benchmark_model`. Can you share how do you get these messages?


Thanks a lot!

**Comment by Maratyszcza** [comment]
I just build with Bazel flags `--define xnn_log_level=info --define xnn_log_to_stdio=true`. It is supposed to work.

Hexagon delegate and built-in CPU kernels don't fail because they don't fully validate parameters of the model.

**Comment by Janus-Shiau** [comment]
@Maratyszcza @ngzhian ngzhian
It seems that `--define xnn_log_level=info --define xnn_log_to_stdio=true` works for tensorflow 2.11.0 release. I could get detailed debug message from XNNPACK.

But it doesn't work for tensorflow 2.7.0 release, do you have any idea about how to build tflite get detailed XNNPACK debug message in tensorflow 2.7?

Thank you very much for your time on this matter!
